src/librustc_mir/interpret/place.rs

```diff
@@ -348,8 +348,13 @@ where
                 offsets[usize::try_from(field).unwrap()],
             layout::FieldPlacement::Array { stride, .. } => {
                 let len = base.len(self)?;
-                assert!(field < len, "Tried to access element {} of array/slice with length {}",
-                    field, len);
+                if field >= len {
```

Comment:
hmm.. so the root problem here is that promotion only promotes the access code but not the assertions. cc @eddyb @RalfJung 	should we paper over this problem by doing what this PR does? I think it is the best solution, as everything else will just make promotion more complex.

Reply:
Turning this from an assertion to an `EvalError` is fine for me.

Please add a comment though saying that this can be violated because this runs during promotion on code where the type system has not yet ensured that such things don't happen.

Reply:
Done